R/modified_update.R

```diff
@@ -537,105 +536,62 @@ mu_calculate_rate <- function(name, dat, window, n_years) {
-    
-    dat <- mu_calculate_rate("deaths", dat, window, n_years)
-    dat <- mu_calculate_rate("cases", dat, window, n_years)
-    
-    dat$target_pop_estimated <- dat$fvps / dat$coverage_old
-    dat$target_pop_estimated[dat$fvps > 0 & is_blank(dat$coverage_old)] <- NA
-    
-    dat$target_pop_estimated_avg <-
-      roll_mean_by(dat$target_pop_estimated, window, n_years)
-    
-    
-    dat$target_pop_given <- dat$coverage_target
-    dat$target_pop_given_new <- dat$coverage_target_new
-    
-    
-    ## We also want new fvps which we compute as coverage * target pop
-    v <- "coverage_target_new"
-    dat$fvps_new <- dat$coverage_new * dat[[v]]
-    
-    ## At the moment we can do nothing for MHL, TUV and XK, because we have no routine pop for them
-    i <- !is_blank(dat$coverage_new) & is_blank(dat$fvps_new) & dat$year < 2031
-    if (any(i)) {
-      if (x$activity_type == "routine" &&
-          all(dat$country[i] %in% c("MHL", "TUV", "XK"))) {
-        dat$fvps_new[i] <- dat$coverage_new[i] * dat$target_pop_estimated[i]
-      } else if (!all(is_blank(dat$fvps[i])))
-        stop("modified update error")
-    }
-    
-    #For analysis purpose, a touchstone can be 'updated' by itself - assumes equal fvps
-    if(meta$touchstone_mod$touchstone_name == meta$group$touchstone_name[meta$group$index == dat_id[1]])
-      dat$fvps_new <- dat$fvps
-    ## drop excess temporary things
-    dat$.code <- NULL
-    dat <- dat[dat$year <= year_max, ]
-    
-    dat <- rbind(dat, keep)
-  } else {dat <- data}
- dat
+##' Look for introduction year and add to summary output
+##' @title Find year of introduction
+##'
+##' @param dat Data: the list output from the modup
+##'
+##' @param dat_summary Data: the summary output from the modup
+##'
+##' @export
+mu_year_introduction <- function(con, dat, dat_summary){
+  group <- dat$group
+  group$touchstone <- group$touchstone_name
+  meta.group <- c("touchstone", "touchstone_src","modelling_group", "vaccine")
+  meta.group <- unique(group[meta.group])
+  x <- merge(meta.group, dat_summary, by = c("touchstone", "modelling_group", "vaccine"), all.y=TRUE)
+  stopifnot( nrow(x) == nrow(dat_summary) )
+
+  x$touchstone_src[ is.na(x$touchstone_src) ] <- dat$touchstone_mod$id
+
+  coverage.touchstone <- as.list(c(unique(x$touchstone_src)))
+
+  y <- do.call(rbind, lapply(coverage.touchstone, function(i) find_year_introduction(con, i)))
+
+  z <- merge(x, y, by=c("touchstone_src", "activity_type", "vaccine", "country"), all.x = TRUE)
+
+  z
+}
+
+year_intro <-function(a){
+  cols <- c("activity_type", "vaccine", "country", "year", "coverage")
+  cols2 <- c("activity_type", "vaccine", "country", "year_intro")
+
+  i <- is.na(match(colnames(a), cols))
+  if(any(i)) stop("Missing column.")
+
+  a$year_intro <- NA
+  i <- !is.na(a$coverage) & a$coverage != 0.
+  if(any(i)){
+    a$year_intro <- min(a$year[i], na.rm = TRUE)
+  }
+  a <- unique(a[cols2])
+}
+
+find_year_introduction <- function(con, touchstone){
```

Comment:
consider doing more in SQL

```r
d3 <- DBI::dbGetQuery(con, "SELECT DISTINCT
                               activity_type, vaccine, country, MIN(year) AS year_intro
                               FROM coverage
                               JOIN coverage_set
                               ON coverage_set.id = coverage.coverage_set
                               WHERE touchstone = $1
                               AND activity_type = 'routine'
                               AND coverage IS NOT NULL
                               AND coverage != 0
                               GROUP BY vaccine, country, activity_type
                               ORDER BY vaccine, country",
                     touchstone)
```

I think this is about right but might differ.  Please check.  With this you will no longer need the `year_intro` function , nor the split step below

Reply:
Some docs: https://www.techonthenet.com/sql/min.php

Reply:
I'm not 100% sure if the `DISTINCT` is needed once the `GROUP BY` is used

Reply:
Verified. It is supper nice. `DISTINCT` is not needed.